### ISSUE TITLE
Convenient handling of Proxy type?

### ISSUE THREAD

**Issue by bluenote10** [ryansolid/solid#34]
When using Typescript, I cannot find a nice way to work with the implicit `Proxy` type. The type inferred from `createState` wraps objects and arrays into `Proxy<T>`. Now if I want to pass that field into another component the compiler complains, for instance:

```js
function MyComponent(props: {data: number[]}) {
  return <div>...</div>
}

function App() {
  const [state, setState] = createState({
    data: [1, 2, 3]
  })
  return <MyComponent data={(state.data)}/>
}
```

Error:
```
src/typed.tsx:36:23 - error TS2740: Type 'Proxy<number[]>' is missing the following properties from type 'number[]': length, pop, push, concat, and 25 more.

36   return <MyComponent data={(state.data)}/>
```
The inferred type of `state.data` is `Proxy<number[]>` which is obviously correct, but since `Proxy` is a private type, I cannot pass `state.data` around or do anything with it without a cast.

As a work-around I can double cast `state.someProp as any as TheUnderlyingType` everywhere, but that makes working with `state` quite tedious (for example `(state.data as any as number[])[i]` instead of just `state.data[i]`).

I'm wondering
- Can we somehow convince Typescript that a `Proxy<T>` really behaves like a `T` (if that's the case)?
- Should the type system rather hide the wrapping for convenience?
- Exporting `Proxy` would at least allow clients to explicitly use `Proxy<T>` in type signatures. This may reduce the number of casts required, but eventually a cast is still necessary when actually doing something with `T`.

**Comment by ryansolid** [comment]
These are great questions. I honestly do not know. Exporting Proxy seems straight forward enough but as you said I think the fact it's a proxy is not that important for the consumer so hiding it sounds intriguing. I have no idea whether we wouldn't just hit another oddity. That seems be an everyday thing with TypeScript.

**Comment by bluenote10** [comment]
Maybe it is possible to define the proxy type interface like that?

```ts
type Proxy<T extends any> = T & {
  get(): any,
  set(): boolean
};
```

This seems to allow treating a `Proxy<T>` as if it was `T`. The following compiles:

```ts
const x: Proxy<number[]> = {} as Proxy<number[]>
const l = x.length
x[0] = 0
```

**Comment by ryansolid** [comment]
Ok, I mean if that does the trick then I'm pretty stoked. The last thing I want is to deal with unnecessary typings. If there are no objections I will change it and export the type from the project.  I have no idea the best way to solve these things, but if that is a reasonable approach I'm all for it.

**Comment by bluenote10** [comment]
I only had a quick look at `state.ts` so I'm not fully sure what it is doing. After reading up what a Proxy is, I'm a bit surprised that it is defined as

```ts
type Proxy<T> = {
  get(): any,
  set(): boolean
}
```

According to the [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) I would say a Proxy is best described as either `type Proxy<T> = T` or `Wrapped<T>`. The `get` and `set` methods only exist on the handler passed into the Proxy constructor, right? I verified that by actually calling `state.data.get()` and `state.data.set()` which should exist according to the type definitions, but both give runtime errors (`TypeError: state.data.set is not a function`).

I'm also confused by the `wrap` function, where the actual Proxy constructor is returned as a `Wrapped<T>`:
```ts
function wrap<T extends StateNode>(value: T): Wrapped<T> { return value[SPROXY] || (value[SPROXY] = new Proxy(value, proxyTraps)); }
```

How would you describe in natural language what a `Proxy<T>` and `Wrapped<T>` is? Is the wrapping on root level different from the nested wrapping or are they the same?

**Comment by ryansolid** [comment]
Got the Proxy from an example in a section in this doc https://www.typescriptlang.org/docs/handbook/advanced-types.html.

This was how they applied the Proxy to properly forward Types from T.  Wrapped<T> is the outputted Type which represented the Proxy wrapped state, it forwards all the keys from T, and shows that nested Objects are too wrapped, whereas simple primitives return their value. So I'd say Wrapped<T> is more representative of the state object after it has been wrapped in a Proxy. There is no difference in whether they are top level or not. Looking at the code again I would expect you to be using Wrapped<T> not Proxy<T>. And Wrapped<T> is the one that is exported.

**Comment by bluenote10** [comment]
Yes, the difference between the proxy defined in the examples
```ts
type Proxy<T> = {
    get(): T;
    set(value: T): void;
}
```
and the native proxy is that the example hides access behind explicit get/set wrappers, whereas native proxies still behave like `T` itself, but with access traps.

As far as I can see, we could simply remove the proxy type and define `Wrapped` as:
```ts
export type Wrapped<T> = {
  [P in keyof T]: T[P] extends object ? Wrapped<T[P]> : T[P];
} & { _state: T }
```

**Comment by ryansolid** [comment]
Yeah probably my misunderstanding of the example since you define proxies by accessors. If you can confirm this works seems like a simple change.

**Comment by bluenote10** [comment]
Yes, I've been experimenting with this change a bit, and so far it seems to work fine.

**Comment by r0skar** [comment]
I did actually run into the same issue shortly after I proposed the initial change to the `Wrapped` type, however I was so busy at work the last two weeks, I did not investigate it further. Can confirm that @bluenote10 changes seem to work fine so far.

**Comment by ryansolid** [comment]
Fixed in Solid v0.7.6.
